src/engine/LocalVocab.h

```diff
@@ -14,31 +15,34 @@
 #include "global/Id.h"
 #include "parser/LiteralOrIri.h"
 #include "util/BlankNodeManager.h"
+#include "util/Exception.h"
 
-// A class for maintaining a local vocabulary with contiguous (local) IDs. This
-// is meant for words that are not part of the normal vocabulary (constructed
-// from the input data at indexing time).
+// A class for maintaining a local vocabulary, which conceptually is a set of
+// `LiteralOrIri`s that are not part of the original vocabulary (which stems
+// from the input data). The implementation is subtle and quite clever:
 //
-
+// The entries of the local vocabulary are `LocalVocabEntry`s, each of which
+// holds a `LiteralOrIri` and remembers its position in the original vocabulary
+// after it has been computed once.
+//
+// A `LocalVocab` has a primary set of `LocalVocabEntry`s, which can grow
+// dynamically, and a collection of other sets of `LocalVocabEntry`s, which are
+// static. A `LocalVocabEntry` lives exactly as long as it is contained in at
+// least one of the (primary or other) sets of a `LocalVocab`.
 class LocalVocab {
  private:
-  using Entry = LocalVocabEntry;
-  using LiteralOrIri = LocalVocabEntry;
-  // A map of the words in the local vocabulary to their local IDs. This is a
-  // node hash map because we need the addresses of the words (which are of type
-  // `LiteralOrIri`) to remain stable over their lifetime in the hash map
-  // because we hand out pointers to them.
-  using Set = absl::node_hash_set<LiteralOrIri>;
+  // The primary set of `LocalVocabEntry`s, which can grow dynamically.
```

Comment:
The part about stable pointers and hence no `flat_hash_set` etc. is quite important, so I would like to keep it in some way.